### ISSUE TITLE
Implement 'P1831R1 Deprecating volatile: library'

### ISSUE THREAD

**Issue by cjdb** [llvm/llvm-project#100038]

**Comment by jkarns275** [comment]
Working on this

**Comment by d3x0r** [comment]
What do I use instead?
If I have a type, and basically all instances of those could be volatile, then I can no longer make this volatile in the typedef, since functions can't return volatile types anymore.... even though the target is volatile, and is subject to change by outsource forces.

typedef { int count; type members[]; } LIST, volatile PLIST *volatile; PLIST pList; AddItem( &pList, whatever );  in the next instant, pList might change its value, since the address of the list was passed, so another add item in some other thread can end up changing either the content of the point or the value of this pointer itself.

This has nothing to do with 'atomic' just, 'don't optimize reads' or don't use a cached value, but if you need the value, read the value from memory.  This is just 'read always' Writes could probably be cached; but I dunno maybe setting data to something like a parallel port each write might do a trigger of a strobe or something so every write even of the same value should probably also happen...

**Comment by philnik777** [comment]
@d3x0r atomics are exactly the thing you want to use for thread synchronization. Anyways, this is not the place to complain about committee decisions.

**Comment by d3x0r** [comment]
@philnik777 Yes sorry it did include a lot of venting; But a title about the topic I want to discuss, with a link to the relatvent information and nothing else - looked pretty open to me. 

But more importantly this project implemented it - so surely they have have an idea about how to deal with such a change or migration hints.  Atomics have nothing to do with it - there is no synchronization required.

I just need to tell any optimizer to always read the value from memory.

So apparently now I need to define

```
typedef struct LinkBlock volatile* volatile PLIST;
typedef struct LinkBlock volatile* PLIST_rval;
```
so I then have to declare

```
PLIST_rval CreateList( void );
```

and then use it 

```
PLIST list = CreateList();
```

That's some clean looking code. 

and if I pass a PLIST now I have to define the parameter type as a PLIST_rval instead of just PLIST like the originating type... at least it doesn't force me to include a typecast like passing `char**` to `char const *const*const`.

Should have maybe started with deprecating volatile in typedefs.

**Comment by jkarns275** [comment]
The rationale is given [here](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1152r0.html#parret)

**Comment by philnik777** [comment]
@d3x0r This is for tracking the implementation status, not asking questions about how to do something. If you want help transitioning feel free to ask on Discord or Discourse, but not on this issue.
